blog/_posts/2020-11-24-ranluxpp.md

```diff
@@ -0,0 +1,157 @@
+A widely used collection of empirical tests is [TestU01][TestU01] by L'Ecuyer and Simard.
+It has to be noted that the tests can only prove the existence of defects.
+That is, passing all tests does not mean that a PRNG is free of errors.
+
+[TestU01]: http://simul.iro.umontreal.ca/testu01/tu01.html
+
+As with many other design issues, PRNGs need to make a certain trade-off:
+fast generators with a small state will usually fail many statistical tests.
+On the other hand, good generators are typically slower or use a much larger state.
+In fact, most generators will fail one test or another, including the widely used Mersenne Twister.
+Notable exceptions are RANLUX and MIXMAX that pass all currently known tests.
+However, they are rather slow (RANLUX) or require a large state (MIXMAX).
+
+## RANLUX++
+
+[RANLUX++][paper] is an LCG implementation derived from the ideas of RANLUX.
+It aims to provide better performance while maintaining the good properties of RANLUX.
+The generator uses a single 576 bit number as state.
+When invoked, it hands out the next unused bits that the generator remembers as an offset.
+Once all bits are exhausted, the LCG performs a multiplication and a modulo operation to advance the state.
+
+[paper]: https://arxiv.org/pdf/1705.03123.pdf
+
+Arithmetic on 576 bit numbers is way beyond what current hardware offers.
+Instead the state is portably stored as 9 numbers of 64 bits each.
+On this representation, multiplication and the modulo operation must thus be done in software.
+For the multiplication, this is very similar to the way multiplications are taught in high school.
+The modulo operation can take advantage of the known modulus (for details, refer to section 2.1 of the [paper][paper]).
+
+For efficiency reasons, Sibidanov [implemented the routines in x86 assembly][ranluxpp].
+This results in impressive performance, but it is not portable across all architectures supported by ROOT.
+So how much slower would a portable implementation be?
+After implementing RANLUX++ for ROOT, it turns out: not much!
+
+[ranluxpp]: https://github.com/sibidanov/ranluxpp/
+
+The [initial version][initial] added to ROOT takes around 30ns per random `double` sampled.
+For comparison, the implementation by Sibidanov comes in at around 8ns per number.
+This was measured with 10 repetitions of a [microbenchmark][benchmark] with a standard deviation of less than 1%.
+The numbers are from a single core of an AMD Ryzen 9 3900 running CentOS 8.2.2004.
+The default compiler is GCC in version 8.3.1, but I will present more numbers in the following.
+
+[initial]: https://github.com/root-project/root/commit/36cc2a820a2b5a687bb603fcaac5f2c09e574ea2
+[benchmark]: https://github.com/root-project/rootbench/pull/200
+
+## Performance tuning
+
+30ns per number is already quite impressive and twice as fast compared to RANLUX as implemented in `TRandom1` (around 62ns per number).
```

Comment:
I would mention here that a new C Ranlux implementation by M. Luscher (see http://luscher.web.cern.ch/luscher/ranlux/) is also around  30 ns/call (see https://link.springer.com/article/10.1007/s41781-019-0034-3)